src/main/java/org/truffleruby/core/basicobject/BasicObjectNodes.java

```diff
@@ -126,11 +126,16 @@ public boolean equal(DynamicObject a, DynamicObject b) {
             return a == b;
         }
 
-        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "notSameClass(a, b)", "isNotIntLong(a) || isNotIntLong(b)" })
+        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "!sameClass(a, b)", "isNotIntLong(a) || isNotIntLong(b)" })
         public boolean equalIncompatiblePrimitiveTypes(Object a, Object b) {
             return false;
         }
 
+        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "sameClass(a, b)", "isNotIntLongDouble(a) || isNotIntLongDouble(b)" })
```

Comment:
That's what Sulong does.
But anyway that's a larger change, the easier fix for now is `!isBoxedPrimitive(a)`.
If performance is critical we could of course cache on the classes of these arguments.